### ISSUE TITLE
model.fit loss vs. model.evaluate loss and ModelCheckpoint

### ISSUE THREAD

**Issue by mevatron** [tensorflow/similarity#257]
First off, thanks for creating this awesome library! Really enjoying using it.

I've noticed an issue that popped up recently, and I'm not sure if it's related to BatchNorm/Regularization, but wanted to make sure that I'm saving the model correctly.

My current setup involves the following TensorFlow versions:

```
TensorFlow: 2.7.0
TensorFlow Similarity: 0.15.8
```

I am using `ModelCheckpoint` to save the model when the val_loss improves:

```
model = tfsim.architectures.EfficientNetSim(
    input_shape=train_sampler.example_shape,
    pooling="gem",  # Can change to use `gem` -> GeneralizedMeanPooling2D
    gem_p=3.0  # Increase the contrast between activations in the feature map.
)

loss = tfsim.losses.CircleLoss(gamma=256)
lr = 1e-4
model.compile(optimizer=tf.keras.optimizers.Adam(lr), loss=loss)

patience_level = max(10, epochs // 5)
early_stopping = tf.keras.callbacks.EarlyStopping(patience=patience_level, verbose=1)
model_checkpoint = tf.keras.callbacks.ModelCheckpoint(model_name, save_best_only=True, verbose=1)
history = model.fit(
    train_sampler,
    epochs=epochs,
    steps_per_epoch=train_steps,
    validation_data=val_sampler,
    validation_steps=val_steps,
    callbacks=[early_stopping, model_checkpoint]
)
 ```
 
 When saving the model I get the following output:
 
```
Epoch 00002: val_loss improved from 93.36356 to 92.38375, saving model to mymodel
WARNING:absl:Found untraced functions such as dense_layer_call_fn, dense_layer_call_and_return_conditional_losses, dense_layer_call_fn, dense_layer_call_and_return_conditional_losses, dense_layer_call_and_return_conditional_losses while saving (showing 5 of 5). These functions will not be directly callable after loading.
/home/wlucas/PycharmProjects/matchernet-training/venv/lib/python3.8/site-packages/keras/engine/functional.py:1410: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.
  layer_config = serialize_layer_fn(layer)
/home/wlucas/PycharmProjects/matchernet-training/venv/lib/python3.8/site-packages/keras/saving/saved_model/layer_serialization.py:112: CustomMaskWarning: Custom mask layers require a config and must override get_config. When loading, the custom mask layer must be passed to the custom_objects argument.
  return generic_utils.serialize_keras_object(obj)
49/49 [==============================] - 260s 5s/step - loss: 86.2540 - val_loss: 92.3838
```

I get the following `model.evaluate` losses, which seemed pretty far off from the `model.fit`  epoch losses:

```
Without load_model (using model in memory immediately after fit):
model.evaluate for val_loss: 155.82171630859375
model.evaluate for train_loss: 144.23423767089844

With load_model:
model.evaluate for val_loss: 155.82171630859375
model.evaluate for train_loss: 144.23423767089844

Random model untrained model just for comparison:
model.evaluate for val_loss: 181.814697265625
model.evaluate for train_loss: 181.07394409179688
```

Is this the correct way to save and load the model from disk? I noticed there was a `.save` method on the `SimilarityModel` class, but I don't think `ModelCheckpoint` is calling that. Are those warnings during the checkpoint anything to worry about here? It seems like the model load worked since the evaluate in-memory vs. load_model scores are identical, but they are much higher than I would have expected. This is a pretty small dataset, so maybe it is something with the BatchNorm layers not being initialized properly in evaluate?

Any tips would be greatly appreciated!

**Comment by mevatron** [comment]
Silly me, had a copy-pasta error of evaluating directly on `val_x` and `val_y` from another classifier project, looks like the this produces much more reasonable outputs:

```
model.evaluate(val_sampler, steps=val_steps)
model.evaluate for val_loss: 88.4561996459961
```

That is a bit lower than the `model.fit` loss, but I think that's due to the BatchNorm/Regularization layers?

**Comment by owenvallis** [comment]
Glad to you figured it out and the evaluate loss look much better now. I think the 88.45 vs 92.38 difference seems reasonable.
